### PR TITLE
feat: refactoring steve cache

### PR DESCRIPTION
**Pull request by CeerDecy**
#### What this PR does / why we need it:

feat: refactoring steve cache

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/all?id=574417&iterationID=-1&type=TASK)

#### Specified Reviewers:

/assign @iutx 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |      refactoring steve cache       |
| 🇨🇳 中文    |     使用gcache作为Steve客户端的缓存     |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: cmd/erda-server/bootstrap.yaml [hunk 1]
```diff
@@ -368,7 +368,9 @@ erda.dop.rule.action.dingtalkworknotice: { }
 #erda.core.org: {}
 
 ########### cmp
-cmp: { }
+cmp:
+  cache_ttl: "10m"
+  cache_size: 64
```
Comment (iutx) [inline cmd/erda-server/bootstrap.yaml:373]: Add environment variable extensions.

File: internal/apps/cmp/steve/aggregator.go [hunk 2]
```diff
@@ -52,50 +52,64 @@ type group struct {
...
-	a.init()
+
+	a.server = gcache.New(size).Expiration(ttl).LoaderFunc(a.loadFunc).LRU().Build()
 	go a.watchClusters(ctx)
 	return a
 }
 
+func (a *Aggregator) loadFunc(key any) (any, error) {
+	logrus.Infof("load cluster: %s", key.(string))
+	ctx := transport.WithHeader(a.Ctx, metadata.New(map[string]string{httputil.InternalHeader: "true"}))
```
Comment (iutx) [inline internal/apps/cmp/steve/aggregator.go:74]: Use built-in method.

File: internal/apps/cmp/steve/aggregator.go [hunk 3]
```diff
@@ -52,50 +52,64 @@ type group struct {
...
+	cluster, err := a.clusterSvc.GetCluster(ctx, &clusterpb.GetClusterRequest{IdOrName: key.(string)})
```
Comment (iutx) [inline internal/apps/cmp/steve/aggregator.go:75]: Assert and intercept at the function beginning.

File: internal/apps/cmp/steve/aggregator.go [hunk 4]
```diff
@@ -104,12 +118,13 @@ func (a *Aggregator) IsServerReady(clusterName string) bool {
 
 // HasAccess set schemas for apiOp and check access for user in apiOp
 func (a *Aggregator) HasAccess(clusterName string, apiOp *types.APIRequest, verb string) (bool, error) {
-	g, ok := a.servers.Load(clusterName)
-	if !ok {
+	//g, ok := a.servers.Load(clusterName)
```
Comment (iutx) [inline internal/apps/cmp/steve/aggregator.go:121]: Remove unuse code.

File: internal/apps/cmp/steve/aggregator.go [hunk 5]
```diff
@@ -191,67 +206,61 @@ func (a *Aggregator) listClusterByType(types ...string) ([]*clusterpb.ClusterInf
...
+	}
+	go a.prepareSteveServer(clusterInfo)
+}
 
-		g := &group{
-			ready:  true,
-			server: server,
-			cancel: cancel,
+// prepareSteveServer creates steve server for a cluster.
+func (a *Aggregator) prepareSteveServer(clusterInfo *clusterpb.ClusterInfo) {
```
Comment (iutx) [inline internal/apps/cmp/steve/aggregator.go:229]: clusterInfo != nil

File: internal/apps/cmp/steve/aggregator.go [hunk 6]
```diff
@@ -191,67 +206,61 @@ func (a *Aggregator) listClusterByType(types ...string) ([]*clusterpb.ClusterInf
...
+	logrus.Infof("creating predefined resource for cluster %s", clusterInfo.Name)
+	if err := a.createPredefinedResource(clusterInfo.Name); err != nil {
+		logrus.Infof("failed to create predefined resource for cluster %s, %v. Skip starting steve server",
+			clusterInfo.Name, err)
+		//a.server.Remove(clusterInfo.Name)
```
Comment (iutx) [inline internal/apps/cmp/steve/aggregator.go:234]: Remove unuse code.

File: internal/apps/cmp/steve/aggregator.go [hunk 7]
```diff
@@ -325,17 +334,22 @@ func (a *Aggregator) insureSystemNamespace(client *k8sclient.K8sClient) error {
...
+		return
+	}
 
 	group, _ := g.(*group)
 	if group.ready {
 		group.cancel()
 	}
-	a.servers.Delete(clusterName)
+	a.server.Remove(clusterName)
+	//a.servers.Delete(clusterName)
```
Comment (iutx) [inline internal/apps/cmp/steve/aggregator.go:352]: Remove unuse code.

File: internal/apps/cmp/steve/aggregator.go [hunk 8]
```diff
@@ -350,8 +364,9 @@ func (a *Aggregator) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	s, ok := a.servers.Load(clusterName)
-	if !ok {
+	s, err := a.server.Get(clusterName)
+	//s, ok := a.servers.Load(clusterName)
```
Comment (iutx) [inline internal/apps/cmp/steve/aggregator.go:368]: Remove unuse code.

File: internal/apps/cmp/steve/aggregator.go [hunk 9]
```diff
@@ -191,67 +206,61 @@ func (a *Aggregator) listClusterByType(types ...string) ([]*clusterpb.ClusterInf
 	return result, nil
 }
 
-func (a *Aggregator) init() {
-	clusters, err := a.listClusterByType("k8s", "edas")
```
Comment (iutx) [inline internal/apps/cmp/steve/aggregator.go:195]: Please add this part of logic during load function.

File: internal/apps/cmp/steve/aggregator.go [hunk 10]
```diff
@@ -52,50 +52,64 @@ type group struct {
...
-	})
+	}
 	return clustersNames
 }
 
 // ListClusters list ready and unready clusters in steveAggregator
 func (a *Aggregator) ListClusters() (ready, unready []string) {
-	a.servers.Range(func(key, value interface{}) bool {
-		g := value.(*group)
+	servers := a.server.GetALL(false)
```
Comment (iutx) [inline internal/apps/cmp/steve/aggregator.go:98]: Is `GetALL` params still false?

**Review by iutx — CHANGES_REQUESTED** [review]
@CeerDecy

**Review by iutx — DISMISSED** [review]
/LGTM
PTAL :) @sfwn

**Comment by iutx** [comment]
@sfwn PTAL

File: cmd/erda-server/bootstrap.yaml [hunk 11]
```diff
@@ -368,7 +368,9 @@ erda.dop.rule.action.dingtalkworknotice: { }
 #erda.core.org: {}
 
 ########### cmp
-cmp: { }
+cmp:
+  cache_ttl: ${CMP_STEVE_CACHE_TTL:10m}
+  cache_size: ${CMP_STEVE_CACHE_SIZE:64}
```
Comment (sfwn) [inline cmd/erda-server/bootstrap.yaml:373]: Note the unit so that others can adjust easily.

File: internal/apps/cmp/steve/aggregator.go [hunk 12]
```diff
@@ -52,50 +54,68 @@ type group struct {
...
-	})
+	}
 	return
 }
 
 func (a *Aggregator) IsServerReady(clusterName string) bool {
-	s, ok := a.servers.Load(clusterName)
-	if !ok {
+	s, err := a.server.Get(clusterName)
+	if err != nil {
```
Comment (sfwn) [inline internal/apps/cmp/steve/aggregator.go:118]: log error is better.

File: internal/apps/cmp/steve/aggregator.go [hunk 13]
```diff
@@ -104,12 +124,12 @@ func (a *Aggregator) IsServerReady(clusterName string) bool {
 
 // HasAccess set schemas for apiOp and check access for user in apiOp
 func (a *Aggregator) HasAccess(clusterName string, apiOp *types.APIRequest, verb string) (bool, error) {
-	g, ok := a.servers.Load(clusterName)
-	if !ok {
+	item, err := a.server.Get(clusterName)
+	if err != nil {
 		return false, errors.Errorf("steve server not found for cluster %s", clusterName)
```
Comment (sfwn) [inline internal/apps/cmp/steve/aggregator.go:129]: missing error detail

File: internal/apps/cmp/steve/aggregator.go [hunk 14]
```diff
@@ -404,7 +442,7 @@ func (a *Aggregator) Serve(clusterName string, apiOp *types.APIRequest) error {
 
 		logrus.Infof("steve for cluster %s not exist, starting a new server", clusterInfo.Name)
 		a.Add(clusterInfo)
-		if s, ok = a.servers.Load(clusterInfo.Name); !ok {
+		if s, err = a.server.Get(clusterInfo.Name); err != nil {
 			return apierrors2.ErrInvoke.InternalError(errors.Errorf("failed to start steve server for cluster %s", clusterInfo.Name))
```
Comment (sfwn) [inline internal/apps/cmp/steve/aggregator.go:446]: missing error detail too

**Review by sfwn — COMMENTED** [review]
Handle errors in more detail.

**Review by sfwn — APPROVED** [review]

**Pull request merged**